File: src/Web/Client/Services/OrganizationsService.cs
using Flurl;
using MudBlazor;
using Shared.Dtos;
using Shared.Enums;
using Shared.ViewModels;

namespace Web.Client.Services;

public class OrganizationsService : ApiService
{
    public OrganizationsService(HttpClient httpClient, ISnackbar snackbar)
        : base(httpClient, snackbar)
    {
        
    }

    public async Task<OrganizationMembersVM?> GetMembers(Guid organizationId)
        => await Get<OrganizationMembersVM>($"organizations/{organizationId}/members");

    public async Task<OrganizationInvitationsVM?> GetInvitations(Guid organizationId, Pagination pagination)
    {
        var url = $"organizations/{organizationId}/invitations"
            .SetQueryParam("pageNumber", pagination.PageNumber)
            .SetQueryParam("itemsPerPage", pagination.ItemsPerPage)
            .ToString();
        return await Get<OrganizationInvitationsVM>(url);
    }

    public async Task<RolesVM<OrganizationPermissions>?> GetRoles(Guid organizationId)
        => await Get<RolesVM<OrganizationPermissions>>($"organizations/{organizationId}/roles");

    public async Task<bool> Create(CreateOrganizationDto model)
        => await Post("organizations", model);

    public async Task<bool> SendInvitation(Guid organizationId, CreateOrganizationInvitationDto model)
        => await Post($"organizations/{organizationId}/invitations", model);

    public async Task<bool> CancelInvitation(Guid organizationId, Guid invitationId)
        => await Post($"organizations/invitations/{invitationId}/cancel", 
            headers: Headers.From(("OrganizationId", organizationId.ToString())));

    public async Task<bool> AcceptInvitation(Guid invitationId)
        => await Post($"organizations/invitations/{invitationId}/accept");

    public async Task<bool> DeclineInvitation(Guid invitationId)
        => await Post($"organizations/invitations/{invitationId}/decline");

    public async Task<OrganizationsForUserVM?> GetForUser()
        => await Get<OrganizationsForUserVM>("organizations/user");

    public async Task<UserOrganizationInvitationsVM?> GetInvitationsForUser()
        => await Get<UserOrganizationInvitationsVM>("organizations/invitations/user");

    public async Task<bool> RemoveMember(Guid organizationId, Guid memberId)
        => await Post($"organizations/{organizationId}/members/{memberId}/remove");

    public async Task<bool> UpdateMemberRole(Guid organizationId, Guid memberId, UpdateMemberRoleDto model)
        => await Post($"organizations/{organizationId}/members/{memberId}/update-role", model);

    public async Task<bool> UpdateRoleName(Guid organizationId, Guid roleId, UpdateRoleNameDto model)
        => await Post($"organizations/{organizationId}/roles/{roleId}/update-name", model);

    public async Task<bool> UpdateRolePermissions(Guid organizationId, Guid roleId, UpdateRolePermissionsDto<OrganizationPermissions> model)
        => await Post($"organizations/{organizationId}/roles/{roleId}/update-permissions", model);

    public async Task<bool> CreateRole(Guid organizationId, CreateRoleDto<OrganizationPermissions> model)
        => await Post($"organizations/{organizationId}/roles", model);

    public async Task<bool> DeleteRole(Guid organizationId, Guid roleId)
        => await Post($"organizations/{organizationId}/roles/{roleId}/delete");

    public async Task<OrganizationNavigationVM?> GetNavData(Guid organizationId)
        => await Get<OrganizationNavigationVM>($"organizations/{organizationId}/nav-data");
}

File: src/Web/Client/composables/organizationsService.ts
import type { CreateOrganizationDto, CreateOrganizationInvitationDto, UpdateOrganizationNameDto } from "~/types/dtos/organizations";
import type { CreateRoleDto, DeleteRoleDto, RemoveMemberDto, UpdateMemberRoleDto, UpdateRoleNameDto, UpdateRolePermissionsDto } from '~/types/dtos/shared';
import type { OrganizationRolesVM, OrganizationMembersVM, OrganizationNavDataVM, OrganizationsVM, OrganizationInvitationsVM, 
    UserOrganizationInvitationsVM, OrganizationSettingsVM, UserOrganizationPermissionsVM } from "~/types/viewModels/organizations";

export const useOrganizationsService = () => {
    const api = useApi();

    return {
        async getOrganizations() {
            return await api.sendGetRequest<OrganizationsVM>('organizations');
        },
        async getNavData(id: string) {
            return await api.sendGetRequest<OrganizationNavDataVM>(`organizations/${id}/nav-data`);
        },
        async createOrganization(model: CreateOrganizationDto) {
            await api.sendPostRequest('organizations', model);
        },
        async getMembers(id: string) {
            return await api.sendGetRequest<OrganizationMembersVM>(`organizations/${id}/members`);
        },
        async getRoles(id: string) {
            return await api.sendGetRequest<OrganizationRolesVM>(`organizations/${id}/roles`);
        },
        async createInvitation(id: string, model: CreateOrganizationInvitationDto) {
            await api.sendPostRequest(`organizations/${id}/invitations`, model);
        },
        async getInvitations(id: string) {
            return await api.sendGetRequest<OrganizationInvitationsVM>(`organizations/${id}/invitations`);
        },
        async updateMemberRole(id: string, model: UpdateMemberRoleDto) {
            await api.sendPostRequest(`organizations/${id}/members/role`, model);
        },
        async getUserInvitations() {
            return await api.sendGetRequest<UserOrganizationInvitationsVM>('organizations/invitations');
        },
        async acceptInvitation(invitationId: string) {
            await api.sendPostRequest(`organizations/invitations/${invitationId}/accept`, '');
        },
        async declineInvitation(invitationId: string) {
            await api.sendPostRequest(`organizations/invitations/${invitationId}/decline`, '');
        },
        async cancelInvitation(id: string, invitationId: string) {
            await api.sendPostRequest(`organizations/invitations/${invitationId}/cancel`, '',  { 'OrganizationId': id });
        },
        async createRole(id: string, model: CreateRoleDto) {
            await api.sendPostRequest(`organizations/${id}/roles`, model);
        },
        async updateRolePermissions(id: string, model: UpdateRolePermissionsDto) {
            await api.sendPostRequest(`organizations/${id}/roles/permissions`, model);
        },
        async updateRoleName(id: string, model: UpdateRoleNameDto) {
            await api.sendPostRequest(`organizations/${id}/roles/name`, model);
        },
        async deleteRole(id: string, model: DeleteRoleDto) {
            await api.sendPostRequest(`organizations/${id}/roles/delete`, model);
        },
        async removeMember(id: string, model: RemoveMemberDto) {
            await api.sendPostRequest(`organizations/${id}/members/remove`, model);
        },
        async getSettings(id: string) {
            return await api.sendGetRequest<OrganizationSettingsVM>(`organizations/${id}/settings`);
        },
        async updateName(id: string, model: UpdateOrganizationNameDto) {
            await api.sendPostRequest(`organizations/${id}/name`, model);
        },
        async deleteOrganization(id: string) {
            await api.sendPostRequest(`organizations/${id}/delete`, '');
        },
        async getUserPermissions(id: string) {
            return await api.sendGetRequest<UserOrganizationPermissionsVM>(`organizations/${id}/permissions`);
        }
    }
}